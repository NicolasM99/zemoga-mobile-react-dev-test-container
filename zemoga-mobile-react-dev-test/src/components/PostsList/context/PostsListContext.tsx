import React, { FC, useState, createContext, useContext, useEffect } from 'react';

import { generalAction } from 'src/redux/actions';
import { GET_POSTS } from 'src/redux/actionTypes';
import { useRedux } from 'src/redux/hooks/useRedux';
import { POSTS_API } from 'src/redux/paths';

import { PostsListProviderProps, PostsListContextType, IPost } from '../@types/postListContext';

const PostsListContext = createContext<PostsListContextType | null>(null);

const PostsListProvider: FC<PostsListProviderProps> = ({ children }: PostsListProviderProps) => {
  const [posts, setPosts] = useState<IPost[] | []>([]);
  const [postsToDelete, setPostsToDelete] = useState<IPost[] | []>([]);
  const [isDeletingItems, setIsDeletingItems] = useState(false);
  const [dispatch, { posts: postsStore }, status] = useRedux(['posts']);
  const [isLoading, setIsLoading] = useState<boolean>(true);

  useEffect(() => {
    if (postsStore) setPosts([...postsStore]);
  }, [postsStore]);

  useEffect(() => {
    if (status) setIsLoading(false);
  }, [status]);

  useEffect(() => {
    if (isLoading) setIsLoading(false);
  }, [isLoading]);

  useEffect(() => {
    dispatch(generalAction({ actionType: GET_POSTS, api: POSTS_API }));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  const valueObj = {
    posts,
    setPosts,
    isDeletingItems,
    setIsDeletingItems,
    postsToDelete,
    setPostsToDelete,
    isLoading,
    setIsLoading
  };
  return <PostsListContext.Provider value={valueObj}>{children}</PostsListContext.Provider>;
};

const usePostsListContext = () => useContext(PostsListContext);

export { PostsListContext, PostsListProvider, usePostsListContext };
