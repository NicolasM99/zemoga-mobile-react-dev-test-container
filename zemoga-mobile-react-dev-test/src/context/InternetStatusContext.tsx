import React, { createContext, FC, useContext, useEffect, useState } from 'react';

import NetInfo from '@react-native-community/netinfo';

import { IInternetStatusContext, InternetStatusContextType } from './@types/internetStatusContext';

const InternetStatusContext = createContext<InternetStatusContextType | null>(null);

const InternetStatusProvider: FC<IInternetStatusContext> = ({
  children
}: IInternetStatusContext) => {
  const [internetStatus, setInternetStatus] = useState<boolean | null>(true);
  useEffect(() => {
    const unsubscribe = NetInfo.addEventListener((state) => {
      setInternetStatus(state.isConnected);
    });
    return () => {
      unsubscribe();
    };
  }, []);

  const valueObj = {
    internetStatus
  };
  return (
    <InternetStatusContext.Provider value={valueObj}>{children}</InternetStatusContext.Provider>
  );
};

const useInternetStatusContext = () => useContext(InternetStatusContext);

export { InternetStatusContext, InternetStatusProvider, useInternetStatusContext };
