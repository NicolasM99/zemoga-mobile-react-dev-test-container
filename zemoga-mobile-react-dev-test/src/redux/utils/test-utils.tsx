import React, { PropsWithChildren } from 'react';

import { configureStore } from '@reduxjs/toolkit';
import type { PreloadedState } from '@reduxjs/toolkit';
import { render } from '@testing-library/react-native';
import type { RenderOptions } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import { INITIAL_STATE } from '../reducers';
import { persistedReducer, store as reduxStore } from '../store';
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<typeof persistedReducer>;
  store?: typeof reduxStore;
}

export const renderWithStoreProviders = (
  ui: React.ReactElement,
  {
    preloadedState = INITIAL_STATE,
    store = configureStore({ reducer: persistedReducer, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const Wrapper = ({ children }: PropsWithChildren<object>): JSX.Element => (
    <Provider store={store}>{children}</Provider>
  );

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};
