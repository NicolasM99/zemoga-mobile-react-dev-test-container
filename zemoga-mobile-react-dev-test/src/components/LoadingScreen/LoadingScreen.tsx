import React, { FC } from 'react';

import { ActivityIndicator, View, Text } from 'react-native';

import { COLORS } from 'src/constants/theme/colors';
import { InternetStatusContextType } from 'src/context/@types/internetStatusContext';
import { useInternetStatusContext } from 'src/context/InternetStatusContext';

import { PostsListContextType } from '../PostsList/@types/postListContext';
import { usePostsListContext } from '../PostsList/context/PostsListContext';

const LoadingScreen: FC = () => {
  const { isLoading } = usePostsListContext() as PostsListContextType;
  const { internetStatus } = useInternetStatusContext() as InternetStatusContextType;
  if (isLoading && internetStatus)
    return (
      <View
        style={{
          zIndex: 9999,
          width: '100%',
          height: '100%',
          justifyContent: 'center',
          alignItems: 'center',
          position: 'absolute',
          backgroundColor: 'rgba(255,255,255,0.9)',
          top: 0,
          left: 0
        }}
      >
        <ActivityIndicator size="large" color={COLORS.primary} />
        <Text style={{ marginTop: 16 }}> Loading, please wait...</Text>
      </View>
    );
  return <></>;
};

export default LoadingScreen;
