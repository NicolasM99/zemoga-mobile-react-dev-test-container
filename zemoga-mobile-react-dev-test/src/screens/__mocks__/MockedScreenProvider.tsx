import React from 'react';

import { NavigationContainer } from '@react-navigation/native';

import { PostsListProvider } from 'src/components/PostsList/context/PostsListContext';
import { InternetStatusProvider } from 'src/context/InternetStatusContext';

interface MockedScreenProviderProps {
  children: React.ReactNode;
}

const MockedScreenProvider: React.FC<MockedScreenProviderProps> = ({
  children
}: MockedScreenProviderProps) => (
  <NavigationContainer>
    <InternetStatusProvider>
      <PostsListProvider>{children}</PostsListProvider>
    </InternetStatusProvider>
  </NavigationContainer>
);

export default MockedScreenProvider;
