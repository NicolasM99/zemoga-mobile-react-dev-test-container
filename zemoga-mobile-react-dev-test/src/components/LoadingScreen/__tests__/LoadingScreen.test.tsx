import { screen } from '@testing-library/react-native';

import * as PostsListContext from 'src/components/PostsList/context/PostsListContext';
import * as InternetStatusContext from 'src/context/InternetStatusContext';
import { renderWithStoreProviders } from 'src/redux/utils/test-utils';
import MockedScreenProvider from 'src/screens/__mocks__/MockedScreenProvider';

import LoadingScreen from '../LoadingScreen';

describe('<LoadingScreen /> unit tests', () => {
  jest.mock('react-redux', () =>
    jest.fn(() => ({
      useDispatch: () => true
    }))
  );
  jest.spyOn(InternetStatusContext, 'useInternetStatusContext').mockReturnValue({
    internetStatus: true
  });
  jest.spyOn(PostsListContext, 'usePostsListContext').mockReturnValue({
    isLoading: true,
    posts: [],
    setPosts: () => {},
    isDeletingItems: false,
    setIsDeletingItems: () => {},
    postsToDelete: [],
    setPostsToDelete: () => {},
    setIsLoading: () => {}
  });
  it('Should mount correctly the component', () => {
    renderWithStoreProviders(
      <MockedScreenProvider>
        <LoadingScreen />
      </MockedScreenProvider>
    );
    expect(screen).toMatchSnapshot();
  });

  it('Should return empty object if internetStatus is false', () => {
    jest.spyOn(InternetStatusContext, 'useInternetStatusContext').mockReturnValue({
      internetStatus: false
    });
    renderWithStoreProviders(
      <MockedScreenProvider>
        <LoadingScreen />
      </MockedScreenProvider>
    );
    expect(screen).toMatchObject({});
  });

  it('Should return empty object if isLoading is false', () => {
    jest.spyOn(PostsListContext, 'usePostsListContext').mockReturnValue({
      isLoading: false,
      posts: [],
      setPosts: () => {},
      isDeletingItems: false,
      setIsDeletingItems: () => {},
      postsToDelete: [],
      setPostsToDelete: () => {},
      setIsLoading: () => {}
    });
    renderWithStoreProviders(
      <MockedScreenProvider>
        <LoadingScreen />
      </MockedScreenProvider>
    );
    expect(screen).toMatchObject({});
  });
});
