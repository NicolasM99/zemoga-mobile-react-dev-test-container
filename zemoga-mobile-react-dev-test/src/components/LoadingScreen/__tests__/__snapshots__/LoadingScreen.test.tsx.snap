// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoadingScreen /> unit tests Should mount correctly the component 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(255,255,255,0.9)",
      "height": "100%",
      "justifyContent": "center",
      "left": 0,
      "position": "absolute",
      "top": 0,
      "width": "100%",
      "zIndex": 9999,
    }
  }
>
  <ActivityIndicator
    color="#10c64e"
    size="large"
  />
  <Text
    style={
      Object {
        "marginTop": 16,
      }
    }
  >
     Loading, please wait...
  </Text>
</View>
`;
